lib/miq_automation_engine/engine/miq_ae_engine/miq_ae_object_lookup.rb

```diff
@@ -0,0 +1,60 @@
+module MiqAeEngine
+  module MiqAeObjectLookup
+    def initialize_obj_entries
+      @lookup_table = []
+    end
+
+    def add_obj_entry(fqns, klass, instance, object)
+      parts = fqns.split('/')
+      domain = parts.shift
+      namespace = parts.join('/')
+
+      @lookup_table << {:domain    => domain.downcase,
+                        :namespace => namespace.downcase,
+                        :klass     => klass.downcase,
+                        :instance  => instance.downcase,
+                        :object    => object}
+    end
+
+    def find_obj_entry(path)
+      query = path[0] == '/' ? parse_obj_path(path[1..-1]) : parse_obj_path(path)
+
+      entries = @lookup_table.select { |entry| entry[:klass] == query['class_name'] }
+      ientries = entries.select { |entry| instance_match?(query['instance'], entry) }
+      find_best_match(ientries, query)
+    rescue MiqAeException::InvalidPathFormat => err
+      $miq_ae_logger.error(err.message)
+      nil
+    end
+
+    def find_best_match(ientries, query)
+      match = ientries.detect { |entry| fq_match?(query['domain'], query['partial_ns'], entry) }
+      match ||= ientries.detect { |entry| ns_match?(query['fqns'], entry) }
```

Comment:
Again with all of these, you are ultimately doing a Array scanning operation.  It is preferable to construct the lookup table and the sub-tables so that you are doing optimal Hash lookups instead of Array scans.  The operations like fq_match? and ns_match? are very simplistic, so constructing the lookup table with keys like, for fq_match: `[domain, namespace]`

Reply:
@Fryguy Because of domain level overrides and partial matches (just namespaces) the search has to be different. The user can call substituion with fully qualified domain/namespace/class/instance or just namespace/class/instance.

Reply:
The ientries is likely really small at this point because it's already gone through 2 filters, so a scan is probably ok